### DISCUSSION TITLE
Kellnr's docker image doubled in size in last release

### DISCUSSION THREAD

**Discussion by aleksanderkrauze** [kellnr/kellnr#276] [General]
I've upgraded kellnr from `5.2.0` to `5.2.1` and I sadly noticed, that the size of docker image was doubled.

```
$ docker images | grep kellnr
ghcr.io/kellnr/kellnr                        5.2.1                       38aacc10a1c7   3 days ago       2.03GB
ghcr.io/kellnr/kellnr                        5.2.0                       069d62c02192   2 weeks ago      1.01GB
```

I think this was caused by #255. Is there any possibility to trim down some of those newly added dependencies?

**Comment by syphar** [comment]
https://github.com/kellnr/kellnr/pull/255#issuecomment-2031142471

> I understand that the doc build is executed inside the kellnr docker container, so these dependencies are only needed to build docs hosted on kellnr?
> 
> This will only become more and more.
> 
> docs.rs and crater are using [our own `crates-build-env` container](https://github.com/rust-lang/crates-build-env), see also https://docs.rs/about/builds, and the [list of installed packages is long](https://github.com/rust-lang/crates-build-env/blob/master/linux/packages.txt).

**Comment by secana** [comment]
Thanks for reporting. I see a general conflict of interest in what Kellnr wants to achieve. 

1. It should be very easy to setup and use. Just one `docker run` -> That speaks for a build-in doc.rs build
2. It should be small enough to run on low-end hardware -> This speaks for a separat builder

The only option which would solve both is to split Kellnr into three images.

1. The full image with a build in doc.rs builder (what we have now)
2. A minimal Kellnr image without the doc.rs build
3. A builder container which can be used as a sidecar for the 2. to build if needed

With this option, a user can decide if he needs the docs.rs build and if he wants one large image or two smaller one. Unfortunately that is a lot of work to implement.

**Comment by syphar** [comment]
another factor I can imagine: 

if kellnr is used more as a private registry, it's totally fine to just extend the main kellnr docker container with the system dependencies your packages need. 

If you use kellnr more as a mirror of crates.io, you'll need the full image over time.

**Comment by syphar** [comment]
two other things to add because I just had to look it up for another question: 
- we also have a `linux-micro` container which just includes build-essential packages: https://github.com/rust-lang/crates-build-env/blob/master/linux-micro/packages.txt 
- `rustwide` (the library that docs.rs & crater use to run the builds) has an example how a containerized doc-build could work: https://github.com/rust-lang/rustwide/blob/master/examples/docs-builder.rs

**Comment by mjpieters** [comment]
I'd be in favour of moving docs building to a sidecar. You'd need some kind of RPC mechanism to transfer the crate to build the docs for, and to include the necessary source mappings automatically (see my [config.toml thread for notes](https://github.com/kellnr/kellnr/discussions/335)).

Doc builds require a whole different resource profile (CPU, memory, etc.) from a registry, after all. A sidecar container gives devops a much wider range of management options.

**Comment by secana** [comment]
There is a now a "minimal" image without the rust toolchain needed to compile the docs in the image itself.
